Added Pagination Config Option Through Headers.

nearbycoder opened the pull request:
This will allow for pagination information to be past through the
headers on both the request and response. The request should have a
Range-Unit: items and a Content-Range: 0-4. The 0-4 is the amount of
items per page via the current page.

In addition the response will return the Content-Range: 0-4/10, the 0-4
is the range you requested and the 10 will be the total posts.

This change pairs well with the
angular-paginate-anything library found here:
https://github.com/begriffs/angular-paginate-anything

nearbycoder committed:
Added Pagination Config Option Through Headers.

This will allow for pagination information to be past through the
headers on both the request and response. The request should have a
Range-Unit: items and a Content-Range: 0-4. The 0-4 is the amount of
items per page via the current page.

In addition the response will return the Content-Range: 0-4/10, the 0-4
is the range you requested and the 10 will be the total posts.

This change pairs well with the
angular-paginate-anything library found here:
https://github.com/begriffs/angular-paginate-anything

claycarpenter commented on Gemfile.lock:
```diff
@@ -214,16 +215,17 @@ GEM
-      actionpack (>= 3.0)
-      activesupport (>= 3.0)
-      sprockets (>= 2.8, < 4.0)
+    sprockets-rails (3.0.4)
```
I have concerns about the downstream effects of these major version updates. Where the dependency updates necessary?

claycarpenter commented:
This new configuration option probably should also be noted in an update to the project's README.

claycarpenter commented on app/controllers/fuel_controller.rb:
```diff
@@ -14,4 +14,44 @@ def s3_bucket
   end
 
+  def self.paginated_action(options={})
+    before_filter(options) do |controller|
```
Suggest determining all pagination attributes here.

Look for the `config.header_pagination` flag, and if present then attempt to use Range. If not present then look for `params[:page]`. If all pagination configs are left undefined then set current page to `1` and per_page to `nil`; I believe that instructs Kaminari to return everything.

nearbycoder replied:
Good thinking. Less checking within the posts controller.

claycarpenter commented on app/controllers/fuel_controller.rb:
```diff
@@ -14,4 +14,44 @@ def s3_bucket
+      @pagination_current_page = 1
+      @pagination_per_page = nil
+
+      if request.headers['Range-Unit'] == 'items' && request.headers['Range'].present?
```
How does these Range options interact with the existing `config.paginates_per` configuration?

nearbycoder replied:
It directly counteracts any changes to paginates_per, so it is either one or the other.

claycarpenter commented on app/controllers/fuel/posts_controller.rb:
```diff
@@ -3,13 +3,20 @@ class PostsController < FuelController
     layout Fuel.configuration.layout if Fuel.configuration.layout
     before_filter :define_title
+    if Fuel.configuration.header_pagination
+      paginated_action only: [:index]
```
Suggest treating all calls to PostsController#index as pagination-aware calls. See my comments on `fuel_controller.rb`.

claycarpenter commented on app/controllers/fuel/posts_controller.rb:
```diff
@@ -3,13 +3,20 @@ class PostsController < FuelController
     def index
-      @posts = Fuel::Post.recent_published_posts.page(params[:page])
+      if Fuel.configuration.header_pagination
+        @posts = Fuel::Post.recent_published_posts.page(@pagination_current_page).per(@pagination_per_page)
```
We should be able to determine enough during the pre-action filter (see my comments on `fuel_controller.rb`) that we can get away with using only this single line to retrieve Posts for all invocations of the action.

nearbycoder committed:
Refactored to move param/config checking to before_filter

This cleans up the posts_controller. In addition I modified the readme
to include information that was missing about the original pagination
options as well as the new pagination option.

nearbycoder commented:
Ready to be reviewed.

claycarpenter commented on app/controllers/fuel_controller.rb:
```diff
@@ -14,4 +14,44 @@ def s3_bucket
+
+      # Set defaults
+      @pagination_current_page = params[:page] ? params[:page] : 1
+      @pagination_per_page = nil
```
Perhaps I'm reading this wrong, but shouldn't this `pagination_per_page` var be set to `config.paginates_per` when not using the `header_pagination`-style pagination?

claycarpenter commented:
Looks good :+1:
